### PR TITLE
Infer config poutine

### LINKED ISSUE #774: Create a poutine or messenger to configure site['infer']
It would be nice to have a poutine or messenger that could set `site['infer']` properties. Using this we could e.g. simplify the `enum_discrete` logic to first set `site['infer']['enumerate'] = 'sequential'` for all discrete sites, then use existing logic without `site_is_discrete`.

### PR DESCRIPTION
**Pull request by eb8680**
Closes #774 

cc @fritzo

### CODE REVIEW

File: pyro/poutine/infer_config_poutine.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        :param msg: current message at a trace site.
+
+        If self.config_fn is not None, calls self.config_fn on msg
+        and stores the result in msg["infer"].
+
+        Otherwise, implements default sampling behavior
+        with no additional effects.
+        """
+        if self.config_fn is not None:
+            msg["infer"] = msg.get("infer", {})
```
Comment (fritzo) [inline pyro/poutine/infer_config_poutine.py]: nit: You can simplify using `.setdefault()`
```py
msg["infer"].setdefault("enumerate", {}).update(self.config_fn(msg))
```

File: pyro/poutine/infer_config_poutine.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+from __future__ import absolute_import, division, print_function
+
+from .poutine import Messenger, Poutine
+
+
+class InferConfigMessenger(Messenger):
+    """
+    Modifies contents of the infer kwarg at sample sites
+    """
+    def __init__(self, config_fn=None):
```
Comment (fritzo) [inline pyro/poutine/infer_config_poutine.py]: I think there should be no default value for `config_fn`. If we absolutely need a default value, it should be `lambda site: {}`.

File: pyro/poutine/infer_config_poutine.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+        Otherwise, implements default sampling behavior
+        with no additional effects.
+        """
+        if self.config_fn is not None:
+            msg["infer"] = msg.get("infer", {})
+            msg["infer"].update(self.config_fn(msg))
+        return None
+
+    def _pyro_param(self, msg):
```
Comment (fritzo) [inline pyro/poutine/infer_config_poutine.py]: Do param sites not have infer dicts? Is it a bad idea to specify per-site optimizer params in an infer dict?

File: pyro/poutine/infer_config_poutine.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        return None
+
+
+class InferConfigPoutine(Poutine):
+    """
+    Modifies contents of the infer kwarg at sample sites
+    """
+    def __init__(self, fn, config_fn=None):
```
Comment (fritzo) [inline pyro/poutine/infer_config_poutine.py]: ditto: config_fn should not have a default value.

File: pyro/poutine/__init__.py [hunk 5]
```diff
@@ -123,6 +124,20 @@ def condition(fn, data):
...
+def infer_config(fn, config_fn):
+    """
+    :param fn: a stochastic function (callable containing pyro primitive calls)
+    :param config_fn: a callable taking a site and returning an infer dict
+
+    Alias for InferConfigPoutine constructor.
+
+    Given a callable that contains Pyro primitive calls
+    and a callable taking a trace site and returning a dictionary,
+    set the value of the infer kwarg at a sample site to config_fn(site)
```
Comment (fritzo) [inline pyro/poutine/__init__.py]: "updates the value"

File: pyro/poutine/__init__.py [hunk 6]
```diff
@@ -123,6 +124,20 @@ def condition(fn, data):
     return ConditionPoutine(fn, data=data)
 
 
+def infer_config(fn, config_fn):
+    """
+    :param fn: a stochastic function (callable containing pyro primitive calls)
+    :param config_fn: a callable taking a site and returning an infer dict
+
+    Alias for InferConfigPoutine constructor.
```
Comment (fritzo) [inline pyro/poutine/__init__.py]: nit: You can use sphinx notation to link to the class here:
```py
"""
Alias for :class:`~pyro.poutine.infer_config_poutine.InferConfigPoutine` constructor.
"""
```
The `~` tells sphinx to hide the `pyro.poutine_infer_config_poutine.` prefix, and the result is a clickable link in our docs.

**Comment by fritzo** [comment]
I'm looking forward to using this in #776 !

File: pyro/poutine/infer_config_poutine.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        """
+        :param msg: current message at a trace site.
+
+        If self.config_fn is not None, calls self.config_fn on msg
+        and stores the result in msg["infer"].
+
+        Otherwise, implements default sampling behavior
+        with no additional effects.
+        """
+        msg["infer"].setdefault("enumerate", {})
```
Comment (fritzo) [inline pyro/poutine/infer_config_poutine.py]: Sorry, I think the `enumerate` just shouldn't be there. I think you can simply delete this line.

File: tests/poutine/test_poutines.py [hunk 8]
```diff
@@ -592,3 +592,31 @@ def test_trace_compose(self):
...
+
+        self.config_fn = config_fn
+
+    def test_infer_config_sample(self):
+        cfg_model = poutine.infer_config(self.model, self.config_fn)
+
+        tr = poutine.trace(cfg_model).get_trace()
+
+        assert tr.nodes["a"]["infer"] == {"enumerate": "parallel", "blah": True}
+        assert tr.nodes["b"]["infer"] == {"enumerate": {}, "blah": True}
```
Comment (fritzo) [inline tests/poutine/test_poutines.py]: and I think this should simply be `{"blah": True}`

File: tests/poutine/test_poutines.py [hunk 9]
```diff
@@ -592,3 +592,31 @@ def test_trace_compose(self):
...
+
+        def config_fn(site):
+            if site["type"] == "sample":
+                return {"blah": True}
+            else:
+                return {}
+
+        self.config_fn = config_fn
+
+    def test_infer_config_sample(self):
```
Comment (martinjankowiak) [inline tests/poutine/test_poutines.py]: test for composability?

**Pull request merged**